src/server/blocking_controller.cc

```diff
@@ -275,24 +276,29 @@ void BlockingController::AwakeWatched(DbIndex db_index, string_view db_key) {
 }
 
 // Marks the queue as active and notifies the first transaction in the queue.
-void BlockingController::NotifyWatchQueue(std::string_view key, WatchQueueMap* wqm) {
+void BlockingController::NotifyWatchQueue(std::string_view key, WatchQueueMap* wqm,
+                                          const DbContext& context) {
   auto w_it = wqm->find(key);
   CHECK(w_it != wqm->end());
   DVLOG(1) << "Notify WQ: [" << owner_->shard_id() << "] " << key;
   WatchQueue* wq = w_it->second.get();
 
-  DCHECK_EQ(wq->state, WatchQueue::SUSPENDED);
-  wq->state = WatchQueue::ACTIVE;
-
   auto& queue = wq->items;
   ShardId sid = owner_->shard_id();
 
-  do {
-    WatchItem& wi = queue.front();
-    Transaction* head = wi.get();
+  for (auto wi_it = queue.begin(), wi_e = queue.end(); wi_it != wi_e; ++wi_it) {
+    Transaction* head = wi_it->get();
+    // We check may the transaction be notified otherwise move it to the end of the queue
+    if (!wi_it->keyReadyChecker(owner_, context, head, key)) {
+      queue.push_back(std::move(*wi_it));
+      queue.pop_front();
```

Comment:
you pop the first item, then you advance the iterator. looks like you skip elements.
also, you do not create an infinite loop by iterating and pushing to the end at the same time?

Reply:
no infinite loop and skipped elements, but the iterator can be invalidated. Thx